File: java/src/main/java/io/github/ndsev/ndslivemath/Polygon.java
// SPDX-License-Identifier: BSD-3-Clause
package io.github.ndsev.ndslivemath;

import java.util.ArrayList;
import java.util.List;

/**
 * A set of vertices with a topology type and an orientation query.
 *
 * <p>
 * Port of {@code cpp/include/ndsmath/polygon.h}. The C++ class is a template
 * over a vertex container; here it is specialized directly to a list of
 * {@link Wgs84} vertices. Orientation is computed on the raw {@code (lon, lat)}
 * plane (no WGS84 normalization, no antimeridian handling).
 * </p>
 *
 * <p>
 * The base {@link #isValid()} requires at least one vertex; subclasses (e.g.
 * {@link Wgs84Polygon}) may override it.
 * </p>
 */
public class Polygon {

	/** Winding order of a polygon. Integer values match the C++ enum. */
	public enum Orientation {
		/** Clockwise winding ({@code -1}). */
		CLOCKWISE(-1),
		/** Invalid or undefined orientation ({@code 0}). */
		INVALID_ORIENTATION(0),
		/** Counter-clockwise winding ({@code 1}). */
		COUNTERCLOCKWISE(1);

		private final int value;

		Orientation(int value) {
			this.value = value;
		}

		/**
		 * @return the integer value matching the C++ enum
		 */
		public int value() {
			return this.value;
		}
	}

	/** Polygon topology. Integer values match the C++ enum. */
	public enum PolygonType {
		/**
		 * A simple polygon with an arbitrary number of vertices and no holes; the last
		 * vertex connects back to the first ({@code 0}).
		 */
		SIMPLE_POLYGON(0),
		/** A triangle strip ({@code 1}). */
		TRIANGLE_STRIP(1),
		/** A triangle fan ({@code 2}). */
		TRIANGLE_FAN(2),
		/**
		 * A set of independent triangles; three consecutive vertices form one
		 * ({@code 3}).
		 */
		TRIANGLE_LIST(3),
		/** Illegal polygon type, used to signal failure ({@code 4}). */
		UNKNOWN(4);

		private final int value;

		PolygonType(int value) {
			this.value = value;
		}

		/**
		 * @return the integer value matching the C++ enum
		 */
		public int value() {
			return this.value;
		}
	}

	/** The polygon topology type. */
	protected PolygonType polygonType;

	/** The polygon vertices. */
	protected final List<Wgs84> vertices;

	/** Construct an empty polygon of type {@code UNKNOWN}. */
	public Polygon() {
		this(PolygonType.UNKNOWN, null);
	}

	/**
	 * Construct an empty polygon of the given type.
	 *
	 * @param polygonType
	 *            the polygon topology
	 */
	public Polygon(PolygonType polygonType) {
		this(polygonType, null);
	}

	/**
	 * Construct a polygon with an initial vertex list (copied).
	 *
	 * @param polygonType
	 *            the polygon topology
	 * @param vertices
	 *            optional initial vertices (copied); may be {@code null}
	 */
	public Polygon(PolygonType polygonType, List<Wgs84> vertices) {
		this.polygonType = polygonType;
		this.vertices = (vertices != null) ? new ArrayList<>(vertices) : new ArrayList<>();
	}

	/**
	 * Append a single vertex.
	 *
	 * @param position
	 *            the vertex to append
	 */
	public void addVertex(Wgs84 position) {
		this.vertices.add(position);
	}

	/**
	 * Append a list of vertices, in order.
	 *
	 * @param vs
	 *            the vertices to append
	 */
	public void addVertices(List<Wgs84> vs) {
		this.vertices.addAll(vs);
	}

	/**
	 * Array-subscript access to a vertex (no bounds check beyond the list itself,
	 * like C++).
	 *
	 * @param index
	 *            the vertex index
	 * @return the vertex at {@code index}
	 */
	public Wgs84 get(int index) {
		return this.vertices.get(index);
	}

	/**
	 * Replace a vertex.
	 *
	 * @param index
	 *            the vertex index
	 * @param value
	 *            the replacement vertex
	 */
	public void set(int index, Wgs84 value) {
		this.vertices.set(index, value);
	}

	/**
	 * @return the number of vertices
	 */
	public int size() {
		return this.vertices.size();
	}

	/**
	 * @return the polygon type
	 */
	public PolygonType type() {
		return this.polygonType;
	}

	/**
	 * Set the polygon type.
	 *
	 * @param polygonType
	 *            the new polygon topology
	 */
	public void setType(PolygonType polygonType) {
		this.polygonType = polygonType;
	}

	/**
	 * Whether this is a valid polygon.
	 *
	 * <p>
	 * Base implementation: at least one vertex. Overridden by {@link Wgs84Polygon}
	 * to require {@code >= 3}.
	 * </p>
	 *
	 * @return {@code true} if the polygon has at least one vertex
	 */
	public boolean isValid() {
		return !this.vertices.isEmpty();
	}

	/**
	 * @return the (mutable) list of vertices
	 */
	public List<Wgs84> vertices() {
		return this.vertices;
	}

	/**
	 * Compute the winding order via the signed shoelace formula.
	 *
	 * <p>
	 * Only works for {@code SIMPLE_POLYGON} and a single-triangle
	 * {@code TRIANGLE_LIST} (exactly 3 vertices). All other types return
	 * {@code INVALID_ORIENTATION} without computing area. Collinear vertices (zero
	 * area) also return {@code INVALID_ORIENTATION}. Uses raw {@code (lon, lat)}
	 * doubles; no normalization.
	 * </p>
	 *
	 * @return the polygon's orientation
	 */
	public Orientation orientation() {
		if (this.polygonType != PolygonType.SIMPLE_POLYGON
				&& !(this.polygonType == PolygonType.TRIANGLE_LIST && this.vertices.size() == 3)) {
			return Orientation.INVALID_ORIENTATION;
		}

		int n = this.vertices.size();
		double area = 0.0;
		for (int i = 0; i < n; i++) {
			int j = (i + 1 == n) ? 0 : i + 1;
			area += this.vertices.get(i).dx() * this.vertices.get(j).dy();
			area -= this.vertices.get(i).dy() * this.vertices.get(j).dx();
		}

		if (area > 0) {
			return Orientation.COUNTERCLOCKWISE;
		} else if (area < 0) {
			return Orientation.CLOCKWISE;
		} else {
			return Orientation.INVALID_ORIENTATION;
		}
	}
}

File: python/src/ndslive/math/polygon.py
# SPDX-License-Identifier: BSD-3-Clause
"""Generic polygon container with orientation, mirroring the C++ ``Polygon``.

Port of ``cpp/include/ndsmath/polygon.h``. The C++ class is a template over a
vertex container; here it is specialized directly to a list of
:class:`~ndslive.math.wgs84.Wgs84` vertices. Orientation is computed on the raw
``(lon, lat)`` plane (no WGS84 normalization, no antimeridian handling).
"""

from __future__ import annotations

from enum import IntEnum

from .wgs84 import Wgs84


class Orientation(IntEnum):
    """Winding order of a polygon. Integer values match the C++ enum."""

    CLOCKWISE = -1
    INVALID_ORIENTATION = 0
    COUNTERCLOCKWISE = 1


class PolygonType(IntEnum):
    """Polygon topology. Integer values match the C++ enum."""

    #: A simple polygon with an arbitrary number of vertices and no holes; the
    #: last vertex connects back to the first.
    SIMPLE_POLYGON = 0
    #: A triangle strip.
    TRIANGLE_STRIP = 1
    #: A triangle fan.
    TRIANGLE_FAN = 2
    #: A set of independent triangles; three consecutive vertices form one.
    TRIANGLE_LIST = 3
    #: Illegal polygon type, used to signal failure.
    UNKNOWN = 4


class Polygon:
    """A set of vertices with a topology type and orientation query.

    Mirrors the C++ ``Polygon<Vector>`` template. The base
    :meth:`is_valid` requires at least one vertex; subclasses (e.g.
    :class:`~ndslive.math.wgs84_polygon.Wgs84Polygon`) may override it.
    """

    def __init__(
        self,
        polygon_type: PolygonType = PolygonType.UNKNOWN,
        vertices: list[Wgs84] | None = None,
    ) -> None:
        """Construct a polygon.

        Args:
            polygon_type: The polygon topology. Defaults to ``UNKNOWN``.
            vertices: Optional initial vertices (copied).
        """
        self._polygon_type = polygon_type
        self._vertices: list[Wgs84] = list(vertices) if vertices is not None else []

    def add_vertex(self, position: Wgs84) -> None:
        """Append a single vertex."""
        self._vertices.append(position)

    def add_vertices(self, vertices: list[Wgs84]) -> None:
        """Append a list of vertices, in order."""
        self._vertices.extend(vertices)

    def __getitem__(self, index: int) -> Wgs84:
        """Array subscript access to a vertex (no bounds check, like C++)."""
        return self._vertices[index]

    def __setitem__(self, index: int, value: Wgs84) -> None:
        self._vertices[index] = value

    def __len__(self) -> int:
        return len(self._vertices)

    def type(self) -> PolygonType:
        """Get the polygon type."""
        return self._polygon_type

    def set_type(self, polygon_type: PolygonType) -> None:
        """Set the polygon type."""
        self._polygon_type = polygon_type

    def is_valid(self) -> bool:
        """Whether this is a valid polygon.

        Base implementation: at least one vertex. Overridden by
        :class:`~ndslive.math.wgs84_polygon.Wgs84Polygon` to require >= 3.
        """
        return len(self._vertices) > 0

    def vertices(self) -> list[Wgs84]:
        """Get the (mutable) list of vertices."""
        return self._vertices

    def orientation(self) -> Orientation:
        """Compute the winding order via the signed shoelace formula.

        Only works for ``SIMPLE_POLYGON`` and a single-triangle
        ``TRIANGLE_LIST`` (exactly 3 vertices). All other types return
        ``INVALID_ORIENTATION`` without computing area. Collinear vertices
        (zero area) also return ``INVALID_ORIENTATION``.

        Uses raw ``(lon, lat)`` doubles; no normalization.
        """
        if self._polygon_type != PolygonType.SIMPLE_POLYGON and not (
            self._polygon_type == PolygonType.TRIANGLE_LIST and len(self._vertices) == 3
        ):
            return Orientation.INVALID_ORIENTATION

        n = len(self._vertices)
        area = 0.0
        for i in range(n):
            j = i + 1
            if j == n:
                j = 0
            area += self._vertices[i].x * self._vertices[j].y
            area -= self._vertices[i].y * self._vertices[j].x

        if area > 0:
            return Orientation.COUNTERCLOCKWISE
        elif area < 0:
            return Orientation.CLOCKWISE
        else:
            return Orientation.INVALID_ORIENTATION
